platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/PreconsensusEventReplayWorkflow.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.event.preconsensus;
+
+import static com.swirlds.common.formatting.StringFormattingUtils.commaSeparatedNumber;
+import static com.swirlds.common.system.PlatformStatus.READY;
+import static com.swirlds.common.system.PlatformStatus.REPLAYING_EVENTS;
+import static com.swirlds.common.system.PlatformStatus.STARTING_UP;
+import static com.swirlds.common.units.TimeUnit.UNIT_MILLISECONDS;
+import static com.swirlds.logging.LogMarker.EXCEPTION;
+import static com.swirlds.logging.LogMarker.PLATFORM_STATUS;
+import static com.swirlds.logging.LogMarker.STARTUP;
+
+import com.swirlds.common.context.PlatformContext;
+import com.swirlds.common.formatting.UnitFormatter;
+import com.swirlds.common.io.IOIterator;
+import com.swirlds.common.system.PlatformStatus;
+import com.swirlds.common.threading.framework.QueueThread;
+import com.swirlds.common.threading.manager.ThreadManager;
+import com.swirlds.common.time.Time;
+import com.swirlds.logging.payloads.PlatformStatusPayload;
+import com.swirlds.platform.components.EventIntake;
+import com.swirlds.platform.components.EventTaskDispatcher;
+import com.swirlds.platform.components.state.StateManagementComponent;
+import com.swirlds.platform.event.EventIntakeTask;
+import com.swirlds.platform.eventhandling.ConsensusRoundHandler;
+import com.swirlds.platform.internal.EventImpl;
+import com.swirlds.platform.state.signed.ReservedSignedState;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * This class encapsulates the logic for replaying preconsensus events at boot up time.
+ */
+public final class PreconsensusEventReplayWorkflow {
```

Comment:
Would it be practicable to write unit tests for this class?

Reply:
I'll create some unit tests, and ping you when I'm finished.